wyvern/components/ranking_pipeline.py

```diff
@@ -0,0 +1,176 @@
+# -*- coding: utf-8 -*-
+from typing import Any, Generic, List, Optional
+
+from pydantic import BaseModel
+
+from wyvern.components.business_logic.business_logic import (
+    BusinessLogicPipeline,
+    BusinessLogicRequest,
+)
+from wyvern.components.candidates.candidate_logger import CandidateEventLoggingComponent
+from wyvern.components.events.events import LoggedEvent
+from wyvern.components.impressions.impression_logger import (
+    ImpressionEventLoggingComponent,
+    ImpressionEventLoggingRequest,
+)
+from wyvern.components.models.model_component import ModelComponent, ModelInput
+from wyvern.components.pagination.pagination_component import (
+    PaginationComponent,
+    PaginationRequest,
+)
+from wyvern.components.pagination.pagination_fields import PaginationFields
+from wyvern.components.pipeline_component import PipelineComponent
+from wyvern.entities.candidate_entities import ScoredCandidate
+from wyvern.entities.identifier_entities import QueryEntity
+from wyvern.entities.request import BaseWyvernRequest
+from wyvern.event_logging import event_logger
+from wyvern.wyvern_typing import WYVERN_ENTITY
+
+
+class RankingRequest(
+    BaseWyvernRequest,
+    PaginationFields,
+    Generic[WYVERN_ENTITY],
+):
+    query: QueryEntity
+    candidates: List[WYVERN_ENTITY]
+
+
+class ResponseCandidate(BaseModel):
+    candidate_id: str
+    ranked_score: float
+
+
+class RankingResponse(BaseModel):
+    ranked_candidates: List[ResponseCandidate]
+    events: Optional[List[LoggedEvent[Any]]]
+
+
+class RankingPipeline(
+    PipelineComponent[RankingRequest, RankingResponse],
+    Generic[WYVERN_ENTITY],
+):
+    PATH: str = "/ranking"
+
+    def __init__(self, name: Optional[str] = None):
+        self.pagination_component = PaginationComponent[
+            ScoredCandidate[WYVERN_ENTITY]
+        ]()
+        self.ranking_model = self.model
+        self.candidate_logging_component = CandidateEventLoggingComponent[
+            WYVERN_ENTITY,
+            RankingRequest[WYVERN_ENTITY],
+        ]()
+        self.impression_logging_component = ImpressionEventLoggingComponent[
+            WYVERN_ENTITY,
+            RankingRequest[WYVERN_ENTITY],
+        ]()
+
+        upstream_components = [
+            self.pagination_component,
+            self.ranking_model,
+            self.candidate_logging_component,
+            self.impression_logging_component,
+        ]
+        self.business_logic_pipeline: BusinessLogicPipeline
+        if self.business_logic:
+            self.business_logic_pipeline = self.business_logic
```

Comment:
why don't we add the business logic pipeline to upstream components in this case?

Reply:
good catch. it's my oversight. fixed